polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp

```diff
@@ -658,6 +659,47 @@ Value castToIndex(Value val) {
                                             builder.getIndexType(), val);
 }
 
+/// Unroll the given loop \p loop by \p factor.
+void unrollByFactor(LoopLikeOpInterface loop, unsigned factor) {
+  if (factor == 1)
+    return;
+
+  // Lowering affine for loop to scf for loop, as not all affine loops can be
+  // unrolled. Some affine loops (e.g., min expression upper bound) cannot be
+  // unrolled, as the lower bound of the cleanup loop cannot be expressed as an
+  // affine function.
+  if (Operation *loopOp = loop;
+      affine::AffineForOp affineForOp = dyn_cast<affine::AffineForOp>(loopOp)) {
+    OpBuilder builder(affineForOp);
+    Location loc = affineForOp.getLoc();
+    Value lowerBound = lowerAffineLowerBound(affineForOp, builder);
+    Value upperBound = lowerAffineUpperBound(affineForOp, builder);
+    Value step =
+        builder.create<arith::ConstantIndexOp>(loc, affineForOp.getStep());
+    auto scfForOp = builder.create<scf::ForOp>(
+        loc, lowerBound, upperBound, step, affineForOp.getIterOperands());
+
+    // Replace scfForOp body with affineForOp body.
+    scfForOp.getBody()->getParent()->getBlocks().remove(scfForOp.getBody());
+    scfForOp.getRegion().getBlocks().splice(
+        scfForOp.getRegion().end(), affineForOp.getRegion().getBlocks());
+
+    // Replace affine::AffineYieldOp with scf::YieldOp.
+    scfForOp.walk([&](affine::AffineYieldOp yieldOp) {
+      OpBuilder builder(yieldOp);
+      builder.create<scf::YieldOp>(yieldOp.getLoc(), yieldOp.getOperands());
+      yieldOp.erase();
+    });
```

Comment:
```suggestion
    Operation *yieldOp = scfForOp.getRegion().front().getTerminator();
    assert(isa<affine::AffineYieldOp>(yieldOp) && "Expecting affine.yield operation");
    // Same code to replace terminator
```

This may lead to issues. If we have nested `affine.for` loops, we may be replacing their `affine.yield` operations...

Reply:
True, however in this case we know the loop being transformed is the innermost loop because we kicked out outermost loops as part of collecting candidate loops in this transformation. 

We might relax that constraint though so would be best to address this comment or assert ...

Reply:
done.